src/service-worker/service-worker.ts

```diff
@@ -1,37 +1,57 @@
-import {Message, FileRecord} from '../shared/types.js';
-import {endWithSlash} from '../shared/util.js';
-import {expose} from 'comlink';
-import {MESSAGE_TYPES} from '../shared/constants.js';
-
-const swScope = (self as unknown) as ServiceWorkerGlobalScope;
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  HANDSHAKE_RECEIVED,
+  ESTABLISH_HANDSHAKE,
+  ServiceWorkerAPI,
+  FileAPI,
+} from '../shared/worker-api.js';
+import { endWithSlash } from '../shared/util.js';
+import { expose } from 'comlink';
+
+declare var self: ServiceWorkerGlobalScope;
+
+/**
+ * API exposed to the UI thread via Comlink. The static methods on this class
+ * become instance methods on SwControllerAPI.
+ */
+const workerAPI: ServiceWorkerAPI = {
+  setFileAPI(fileAPI: FileAPI, sessionID: string) {
+    fileAPIs.set(sessionID, fileAPI);
+  },
+};
 
-const recieveMessageChannelHandshake = (e: MessageEvent) => {
-  const ports = e.ports;
-  if (ports && ports[0]) {
-    const port = ports[0];
-    port.start();
-    const handshakeReceivedMessage: Message = {
-      type: MESSAGE_TYPES.HANDSHAKE_RECEIVED,
-    };
-    expose(SwController, port);
-    port.postMessage(handshakeReceivedMessage);
+const fileAPIs = new Map<string, FileAPI>();
+const getFile = async (e: FetchEvent, path: string, sessionId: string) => {
+  const fileAPI = fileAPIs.get(sessionId);
+  if (fileAPI) {
+    const file = await fileAPI.getFile(path);
+    if (file) {
+      const headers = file.contentType
+        ? { 'Content-Type': file.contentType }
+        : undefined;
+      return new Response(file.content, { headers });
+    }
   }
+  return fetch(e.request);
 };
 
-interface ResponseParams {
-  content: string;
-  init: ResponseInit;
-}
-
-let fileResponseMap = new Map<string, Map<string, ResponseParams>>();
-
 const onFetch = (e: FetchEvent) => {
-  if (!e.request || !e.request.url || !e.respondWith) {
-    return;
-  }
   const url = new URL(e.request.url);
   const href = url.href;
-  const scope = endWithSlash(swScope.registration.scope);
+  const scope = endWithSlash(self.registration.scope);
```

Comment:
Super congrats on getting TS to accept this as a SW

Reply:
took a bit!